src/global/logger.ts

```diff
@@ -0,0 +1,171 @@
+/* eslint-disable no-console */
+import { config } from './config';
+import { LogLevelIndex, LogLevel } from './logger.type';
+
+interface LogMessage {
+  level: LogLevel;
+  message: string;
+  timestamp: string;
+  data?: any;
+}
+/**
+ * Logging class providing different levels of log
+ */
+class Logger {
+  private static instance: Logger;
+
+  private config: typeof config;
+
+  private constructor() {
+    this.config = config;
+  }
+
+  public static getInstance(): Logger {
+    if (!Logger.instance) {
+      Logger.instance = new Logger();
+    }
+    return Logger.instance;
+  }
+
+  private getTimestamp(): string {
+    return new Date().toISOString();
+  }
+
+  private shouldLog(messageLevel: LogLevelIndex): boolean {
+    const configLevel = this.config.get<string>('logLevel', 'INFO');
+    return messageLevel <= LogLevelIndex[configLevel as LogLevel];
+  }
+
+  private formatMessage(logMessage: LogMessage): string {
+    const { level, message, timestamp, data } = logMessage;
+    let formattedMessage = `[${timestamp}] ${level}: ${message}`;
+
+    if (data) {
+      try {
+        formattedMessage += `\n${JSON.stringify(data, null, 2)}`;
+      } catch (error) {
+        formattedMessage += `\n[JSON.stringify Error/Circular]`;
```

Comment:
Might make sense to propagate the error message since `stringify` could also fail for e.g. bigint values in data.